lib/spack/docs/build_settings.rst

```diff
@@ -398,6 +398,23 @@ under ``all`` are disregarded. For example, with a configuration like this:
 Spack requires ``cmake`` to use ``gcc`` and all other nodes (including ``cmake``
 dependencies) to use ``clang``.
 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Setting requirements for versions that are git commit hashes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you are installing a package that retrieves its source code from a git
+repository, and you want to specify a commit hash from the repository as a
+version requirement, then you need to define an equivalence to a numbered
+version as part of the requirement spec and also mention it as a version
+preference for the package:
+
+.. code-block:: yaml
+
+   packages:
+     openmpi:
+       require: "@full-commit-hash=4.1.5"
+       version: ["4.1.5"]
```

Comment:
Maybe I'm misunderstanding something, but we don't currently require you to specify all this when you just do `spack install foo@abcd1234abcd1234abcd1234abcd1234`.  Why should requirements require more from the user?

Reply:
Note: we *do* require the user to specify a `version:` if 

* you ask for `spack install foo@ abcd1234abcd1234abcd1234abcd1234=1.0` (i.e. a `=numbered-version` at the end)
* and `1.0` isn't mentioned as a `version()` in the `package.py`.

that could be improved, but I carried on that particular constraint here.

Reply:
This is now addressed and the associated tests are updated (plus a preexisting test that was expecting a failure has been updated).